pym/bob/scm/url.py

```diff
@@ -98,24 +96,49 @@ def asScript(self):
 cp -uT {URL} {FILE}
 """
         else:
+            # do command logic inside the shell script to leave it to the backend
             tpl = """
 {HEADER}
 mkdir -p {DIR}
 cd {DIR}
 if [ -e {FILE} ] ; then
-    curl {OPTIONS} -o {FILE} -z {FILE} {URL}
+    if [ -x "$(command -v curl)" ] ; then
+        OPTIONS="-sSgLf"
+        if [ {SKIPCERTCHECK} = True ] ; then
+            OPTIONS+="k"
+        fi
+        curl $OPTIONS -o {FILE} -z {FILE} {URL}
+    fi
 else
     (
         F=$(mktemp)
         trap 'rm -f $F' EXIT
         set -e
-        curl {OPTIONS} -o $F {URL}
+        if [ -x "$(command -v curl)" ] ; then
+            OPTIONS="-sSgLf"
+            if [ {SKIPCERTCHECK} = True ] ; then
+                OPTIONS+="k"
+            fi
+            curl $OPTIONS -o $F {URL}
+        elif [ -x "$(command -v wget)" ] ; then
+            OPTIONS="-q --no-glob"
+            if [ {SKIPCERTCHECK} = True ] ; then
+                OPTIONS+=" --no-check-certificate"
+            fi
+            wget $OPTIONS -O $F {URL}
+        else
+            >&2 echo "\nERROR: Don't know how to download. Please install 'curl'!\n"
+            # cleanup
+            rm -f $F
+            # exit successful to prevent the error print of the script
+            exit 0
```

Comment:
That should be `exit 1`! Otherwise the download will silently fail and break the build because nothing was downloaded as expected.

Reply:
I need to refer to the last two comments...
the #cleanup was necessary in case the exit with '0' has been used since this doesn't involve the trap and the reason was more or less written with the comment (I tried it with '1' before) above the exit.

I thought...
`ERROR: Don't know how to download. Please install 'curl'!

sha1sum: pkg-config-0.29.1.tar.gz: No such file or directory
sha1sum: WARNING: 1 listed file could not be read
Step failed with return status 1; Command: sha1sum -c`
...was sufficient, but you're right, it now complains about the step of checksum verification, which isn't really failing.

Ok, let's summarize using exit with '1' is the correct approach to let the current step fail...but even then I suggest to introduce the error echo. Because this directly tells what was going wrong including a sufficient proposal without reading the printed snippet of the script.